# Olist E-Commerce Analysis

## Context
This project analyzes the [Olist Brazilian E-Commerce dataset](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce) — real transactional data from Brazil's largest department store marketplace, covering 2016–2018.

The dataset contains 99,441 orders across 11 tables: customers, orders, order items, payments, reviews, products, sellers, geolocation, and a separate B2B marketing funnel (8,000 leads → 842 closed deals).

## What this project answers

**Marketing & Sales Funnel (B2B)**
- Lead-to-close conversion rate overall and by acquisition origin
- Which origin converts best and generates most revenue
- Average time from first contact to deal closed

**Customer Profiling (B2C)**
- Average order value by location and product category
- First-timer vs returning customer behavior

**Sellers & Products**
- Best-performing product categories and sellers
- Geographic distribution of revenue

**Operational Performance**
- Delivery time vs estimated delivery
- Review scores vs delivery delay and category

## What this project does NOT answer
No web/clickstream data exists in this dataset — no page views, sessions, bounce rates, or cart events. Those questions are handled in a separate project using the REES46 clickstream dataset.

## Tool stack
PostgreSQL (data storage) → DBeaver (query drafting) → Jupyter + psycopg2 (analysis) → Pandas + Matplotlib/Seaborn (cleaning + visualization)

In [1]:
import psycopg2

conn = psycopg2.connect(
    host="localhost", port=5432,
    database="olit", user="postgres", password="0000"
)
print("Connected.")

Connected.


In [2]:
import glob
import pandas as pd

for file in glob.glob("data/*.csv"):
    df = pd.read_csv(file)
    print(f"--- File: {file} ---")
    print(f"Shape: {df.shape}")
    print(f"Columns: {list(df.columns)}")
    print("First 3 rows:")
    print(df.head(3))
    print("\n")

--- File: data\olist_closed_deals_dataset.csv ---
Shape: (842, 14)
Columns: ['mql_id', 'seller_id', 'sdr_id', 'sr_id', 'won_date', 'business_segment', 'lead_type', 'lead_behaviour_profile', 'has_company', 'has_gtin', 'average_stock', 'business_type', 'declared_product_catalog_size', 'declared_monthly_revenue']
First 3 rows:
                             mql_id                         seller_id  \
0  5420aad7fec3549a85876ba1c529bd84  2c43fb513632d29b3b58df74816f1b06   
1  a555fb36b9368110ede0f043dfc3b9a0  bbb7d7893a450660432ea6652310ebb7   
2  327174d3648a2d047e8940d7d15204ca  612170e34b97004b3ba37eae81836b4c   

                             sdr_id                             sr_id  \
0  a8387c01a09e99ce014107505b92388c  4ef15afb4b2723d8f3d81e51ec7afefe   
1  09285259593c61296eef10c734121d5b  d3d1e91a157ea7f90548eef82f1955e3   
2  b90f87164b5f8c2cfa5c8572834dbe3f  6565aa9ce3178a5caf6171827af3a9ba   

              won_date business_segment      lead_type lead_behaviour_profile  \
0  2018

In [3]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine("postgresql+psycopg2://postgres:0000@localhost:5432/olit")

tables = [
    "customers", "sellers", "products", "product_category_translation",
    "orders", "order_items", "order_payments", "order_reviews",
    "geolocation", "marketing_qualified_leads", "closed_deals"
]

for table in tables:
    df = pd.read_sql(f"SELECT * FROM {table} LIMIT 3", con=engine)
    print(f"\n=== {table} ===")
    print("Shape:", df.shape)
    print(df.dtypes)
    print(df.head(3))


=== customers ===
Shape: (3, 5)
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object
                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  

=== sellers ===
Shape: (3, 4)
seller_id                   str
seller_zip_code_prefix    int64
seller_city                 str
seller_state                str
dtype: object
                          seller_id  seller_z

In [4]:
for table in tables:
    df = pd.read_sql(f"SELECT * FROM {table}", con=engine)
    null_counts = df.isnull().sum()
    null_counts = null_counts[null_counts > 0]
    if not null_counts.empty:
        print(f"\n=== {table} ===")
        print("Shape:", df.shape)
        print(null_counts)


=== products ===
Shape: (32951, 9)
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

=== orders ===
Shape: (99441, 8)
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64

=== order_reviews ===
Shape: (99224, 7)
review_comment_title      87656
review_comment_message    58247
dtype: int64

=== marketing_qualified_leads ===
Shape: (8000, 4)
origin    60
dtype: int64

=== closed_deals ===
Shape: (842, 10)
business_segment            1
lead_type                   6
lead_behaviour_profile    177
business_type              10
dtype: int64


## Data Cleaning

Most nulls in this dataset are structural — missing delivery dates mean orders didn't complete, empty review text is by design. No imputation needed.

One actual cleaning action: `closed_deals` has 4 columns missing ~92% of values (`has_company`, `has_gtin`, `average_stock`, `declared_product_catalog_size`). Keeping them would pollute any segment analysis. Dropped directly in PostgreSQL.

Everything else is documented and left as-is.

In [5]:
from sqlalchemy import text
with engine.connect() as con:
    con.execute(text("""
        ALTER TABLE closed_deals
        DROP COLUMN IF EXISTS has_company,
        DROP COLUMN IF EXISTS has_gtin,
        DROP COLUMN IF EXISTS average_stock,
        DROP COLUMN IF EXISTS declared_product_catalog_size
    """))
    con.commit()
print("Done.")

Done.


## Marketing & Sales Funnel

In [6]:
query = """
select (select count(*) from closed_deals)*100.0/(select count(*) from marketing_qualified_leads) as Lead_to_close_conversion_rate_overall
"""
pd.read_sql(query, con=engine)

,lead_to_close_conversion_rate_overall
0,10.525


In [ ]:
query = """
select 
origin, 
count(*)*100/ (select count(*) from closed_deals) as Percentage_from_closed_deals 
from (select* 
	from closed_deals
	left join marketing_qualified_leads
		using (mql_id)) group by origin order by Percentage_from_closed_deals desc LIMIT 4
"""
pd.read_sql(query, con=engine)

,origin,percentage_from_closed_deals
0,organic_search,32
1,paid_search,23
2,unknown,21
3,social,8


In [ ]:
query = """
SELECT
    origin,
    COUNT(closed_deals.mql_id) * 100.0 / COUNT(*) AS closing_rate
FROM marketing_qualified_leads
LEFT JOIN closed_deals USING (mql_id)
GROUP BY origin
ORDER BY closing_rate DESC
"""
pd.read_sql(query, con=engine)

,origin,closing_rate
0,NaN,23.333333
1,unknown,16.287534
2,paid_search,12.295082
3,organic_search,11.803136
4,direct_traffic,11.222445
5,referral,8.450704
6,social,5.555556
7,display,5.084746
8,other_publicities,4.615385
9,email,3.042596


In [ ]:
query = """
select 
business_segment, 
count(*)*100/(select count(*) from closed_deals) 
     as Percentage_from_closed_deals 
from closed_deals 
group by business_segment 
order by count(*) desc
"""
pd.read_sql(query, con=engine)

,business_segment,percentage_from_closed_deals
0,home_decor,12
1,health_beauty,11
2,car_accessories,9
3,household_utilities,8
4,construction_tools_house_garden,8
5,audio_video_electronics,7
6,computers,4
7,pet,3
8,food_supplement,3
9,food_drink,3


In [ ]:
query = """
SELECT 
    business_segment,
    SUM(declared_monthly_revenue) AS segment_revenue,
    SUM(declared_monthly_revenue) * 100.0 / (SELECT SUM(declared_monthly_revenue) FROM closed_deals) AS revenue_percentage
FROM closed_deals
GROUP BY business_segment
ORDER BY segment_revenue DESC 
LIMIT 3;
"""
pd.read_sql(query, con=engine)

,business_segment,segment_revenue,revenue_percentage
0,construction_tools_house_garden,50695006.0,82.051989
1,phone_mobile,8000000.0,12.948335
2,home_decor,710000.0,1.149165


### Interpretation

The overall lead-to-close conversion rate is **10.5%**.

By origin, `unknown` and `NaN` show the highest closing rates — but since the source is untracked, no actionable decision follows. **Fixing lead-source tracking is a priority**: without it, a meaningful share of conversion performance stays invisible.

Among trackable sources, **organic search (32%)** and **paid search (23%)** together account for 55% of all closed deals and convert above the overall average (11.8% and 12.3% respectively, vs 10.5%). Direct traffic also performs above average. These three channels deserve continued or increased investment.

By business segment, the data does **not** support a segment-level closing rate (no segment field on the leads side) — only volume and revenue of *already closed* deals. This is a tracking gap worth fixing: segment-level conversion would be a strong lever for the business to prioritize lead generation efforts.

What the revenue data does show is stark: **construction_tools_house_garden** generates 82% of total closed-deal revenue while representing only 8% of closed deals by volume. **phone_mobile** follows at a much smaller scale — 12% of revenue from just 1% of deals. By raw deal count, `home_decor` leads (12% of closed deals), but it ranks only 3rd in revenue, contributing roughly 1%.

The implication: the business should concentrate resources on `construction_tools_house_garden`, and secondarily `phone_mobile` — both dwarf every other segment in revenue contribution per deal. That said, this analysis is **revenue-only**. Profit margins by segment are not available in this dataset; without that data, any reallocation of resources risks optimizing for top-line revenue rather than actual profitability. Tracking segment-level cost/margin is a necessary next step before committing budget.

In [ ]:
query = """
select round(EXTRACT(EPOCH from (AVG (won_date - first_contact_date)))/86400) as Average_days_to_close
from (select* 
	from closed_deals
	left join marketing_qualified_leads
		using (mql_id))
"""
pd.read_sql(query, con=engine)

,average_days_to_close
0,49.0


In [ ]:
query = """
select origin,round(EXTRACT(EPOCH from (AVG (won_date - first_contact_date)))/86400) as Average_days_to_close
from (select* 
	from closed_deals
	left join marketing_qualified_leads
		using (mql_id)) group by origin ORDER BY Average_days_to_close ASC
"""
pd.read_sql(query, con=engine)

,origin,average_days_to_close
0,display,11.0
1,other,16.0
2,direct_traffic,32.0
3,referral,33.0
4,other_publicities,40.0
5,unknown,42.0
6,NaN,50.0
7,organic_search,51.0
8,email,53.0
9,paid_search,57.0


In [ ]:
query = """
select business_segment, round(EXTRACT(EPOCH from (AVG (won_date - first_contact_date)))/86400) as Average_days_to_close
from (select* 
	from closed_deals
	left join marketing_qualified_leads
		using (mql_id)) group by business_segment ORDER BY Average_days_to_close DESC
"""
pd.read_sql(query, con=engine)

,business_segment,average_days_to_close
0,NaN,297.0
1,other,254.0
2,perfume,150.0
3,watches,92.0
4,toys,65.0
5,audio_video_electronics,64.0
6,small_appliances,63.0
7,handcrafted,61.0
8,construction_tools_house_garden,61.0
9,music_instruments,59.0


In [ ]:
query = """
SELECT
    origin,
    ROUND(EXTRACT(EPOCH FROM (AVG(won_date - first_contact_date)))/86400) AS average_days_to_close,
    COUNT(*) AS deal_count
FROM closed_deals
LEFT JOIN marketing_qualified_leads USING (mql_id)
WHERE business_segment = 'construction_tools_house_garden'
GROUP BY origin
ORDER BY average_days_to_close DESC
"""
pd.read_sql(query, con=engine)

,origin,average_days_to_close,deal_count
0,NaN,216.0,2
1,email,107.0,2
2,unknown,74.0,19
3,paid_search,69.0,14
4,social,55.0,5
5,organic_search,44.0,19
6,referral,23.0,2
7,direct_traffic,14.0,4
8,display,6.0,2



### Interpretation — Closing time by origin

Our best-converting origins (organic search, paid search, direct traffic) close **above the overall average** of 49 days. This may simply reflect the nature of the business, but less friction in the closing process could likely push the conversion rate even higher — worth investing in.

That said, the highest-revenue segment hadn't yet been examined across origins. Doing so shows: **phone_mobile closes notably faster than average (30 days)**, while **construction_tools_house_garden takes 12 days longer than average (61 days)** despite generating 82% of total closed-deal revenue.

Within `construction_tools_house_garden` specifically, deal volume is concentrated in **organic_search (19 deals, 44 days)**, **unknown (19 deals, 74 days)**, and **paid_search (14 deals, 69 days)**. The `unknown` and `paid_search` origins are both high-volume *and* slow — prime candidates for process improvement.

Given that this segment alone drives the vast majority of revenue, prioritizing its closing process is justified — unless profit margins turn out to be mediocre, since the data only confirms revenue, not profitability. Less friction here translates directly into more money for the business, assuming margins hold up.


In [47]:
query = """
SELECT
    COUNT(*) AS total_sellers,
    COUNT(closed_deals.seller_id) AS sellers_from_closed_deals,
    COUNT(*) - COUNT(closed_deals.seller_id) AS sellers_with_no_deal_record
FROM sellers
LEFT JOIN closed_deals ON sellers.seller_id = closed_deals.seller_id
"""
pd.read_sql(query, con=engine)

,total_sellers,sellers_from_closed_deals,sellers_with_no_deal_record
0,3095,380,2715


### Interpretation — Seller acquisition tracking gap

Of 3,095 total sellers on the platform, only **380 (12.3%)** can be traced back to a recorded closed deal in `closed_deals`. The remaining **2,715 sellers (87.7%)** have no acquisition record at all — no origin, no sales rep, no business segment, no closing timeline.

This is a significant tracking gap. The B2B funnel data (origin, closing time, segment performance) only covers a small fraction of the actual seller base. Any conclusions drawn from the marketing funnel analysis — origin performance, segment revenue, closing time — apply to that 12.3% only, not to how the platform actually acquired most of its sellers.

For the business, this means the current CRM/lead-tracking process is not capturing most seller onboarding. Closing this gap (tracking acquisition for the other 88%) would make the marketing funnel analysis far more representative and useful for resource allocation decisions.

In [48]:
query = """
SELECT
customer_state,
ROUND(SUM(payment_value)) AS revenue_state,
ROUND(AVG(payment_value)) AS avg_revenue_order,
COUNT(*) AS number_orders
FROM customers
JOIN orders USING (customer_id)
JOIN order_payments USING (order_id)
GROUP BY customer_state
ORDER BY revenue_state DESC
"""
df = pd.read_sql(query, con=engine)
df

,customer_state,revenue_state,avg_revenue_order,number_orders
0,SP,5998227.0,138.0,43622
1,RJ,2144380.0,159.0,13527
2,MG,1872257.0,155.0,12102
3,RS,890899.0,157.0,5668
4,PR,811156.0,154.0,5262
5,SC,623086.0,166.0,3754
6,BA,616646.0,171.0,3610
7,DF,355141.0,161.0,2204
8,GO,350092.0,166.0,2112
9,ES,325968.0,155.0,2107


In [49]:
print(df.sort_values("revenue_state", ascending=False).head(5))
print(df.sort_values("avg_revenue_order", ascending=False).head(5))
print(df.sort_values("number_orders", ascending=False).head(5))

  customer_state  revenue_state  avg_revenue_order  number_orders
0             SP      5998227.0              138.0          43622
1             RJ      2144380.0              159.0          13527
2             MG      1872257.0              155.0          12102
3             RS       890899.0              157.0           5668
4             PR       811156.0              154.0           5262
   customer_state  revenue_state  avg_revenue_order  number_orders
15             PB       141546.0              248.0            570
24             AC        19681.0              234.0             84
22             RO        60866.0              233.0            261
25             AP        16263.0              232.0             70
19             AL        96962.0              227.0            427
  customer_state  revenue_state  avg_revenue_order  number_orders
0             SP      5998227.0              138.0          43622
1             RJ      2144380.0              159.0          13527
2   

In [52]:
query = """
SELECT
customer_state,
product_category_name,
ROUND(SUM(payment_value)) AS revenue,
ROUND(AVG(payment_value)) AS avg_revenue_order,
COUNT(*) AS number_orders
FROM customers
JOIN orders USING (customer_id)
JOIN order_items USING (order_id)
JOIN order_payments USING (order_id)
JOIN products USING (product_id)
GROUP BY customer_state, product_category_name
ORDER BY customer_state, revenue DESC
"""
df = pd.read_sql(query, con=engine)
df

,customer_state,product_category_name,revenue,avg_revenue_order,number_orders
0,AC,moveis_decoracao,5261.0,438.0,12
1,AC,esporte_lazer,2072.0,230.0,9
2,AC,beleza_saude,2068.0,295.0,7
3,AC,informatica_acessorios,1860.0,207.0,9
4,AC,relogios_presentes,1574.0,315.0,5
...,...,...,...,...,...
1389,TO,industria_comercio_e_negocios,129.0,129.0,1
1390,TO,casa_construcao,119.0,119.0,1
1391,TO,market_place,81.0,81.0,1
1392,TO,instrumentos_musicais,71.0,71.0,1


In [53]:
df.loc[df.groupby("customer_state")["revenue"].idxmax()].sort_values("revenue", ascending=False)

,customer_state,product_category_name,revenue,avg_revenue_order,number_orders
1282,SP,cama_mesa_banho,765517.0,137.0,5572
931,RJ,cama_mesa_banho,247744.0,139.0,1788
479,MG,cama_mesa_banho,218011.0,155.0,1403
866,PR,moveis_decoracao,114434.0,216.0,530
1107,RS,informatica_acessorios,108921.0,213.0,511
130,BA,beleza_saude,67016.0,170.0,394
1175,SC,informatica_acessorios,66379.0,199.0,333
367,GO,automotivo,57802.0,578.0,100
759,PE,beleza_saude,57240.0,238.0,240
194,CE,beleza_saude,47392.0,272.0,174


### Interpretation — Revenue, AOV, and category by state

The top-revenue states (SP, RJ, MG, RS, PR) have the **lowest average order value**, while low-volume states like PB, AC, RO, AP, AL show the **highest AOV**. SP alone drives ~6M in revenue but averages only 138/order — the lowest AOV on the list.

Category data clarifies this: in nearly every state, the top-revenue category is **cama_mesa_banho** (home/bed/bath) or **beleza_saude** (beauty/health) — both relatively low-ticket, high-frequency categories. SP's top category is cama_mesa_banho with 5,572 orders — high volume of cheaper items, which explains the low AOV despite massive total revenue.

This reframes the earlier read: the high-revenue states aren't underpaying customers — they're driven by **volume in low-ticket categories**. The high-AOV, low-volume states are likely buying fewer, pricier items (or are simply smaller markets with less data to average out outliers).

Business implication: rather than "advertise more in high-AOV states," the better question is category-specific — does SP/RJ/MG have room to grow in higher-ticket categories (informatica_acessorios, moveis_decoracao) the way PR, RS, and SC already show interest in? That's a more actionable growth angle than chasing AOV alone.

In [9]:
query = '''
with order_counts as (
    select customer_unique_id, count(*) as order_numbers
    from customers
    group by customer_unique_id
),
max_orders as (
    select max(order_numbers) as max_n from order_counts
),
levels as (
    select generate_series(1, (select max_n from max_orders)) as level
),
survivors as (
    select l.level, count(oc.customer_unique_id) as customer_count
    from levels l
    left join order_counts oc on oc.order_numbers >= l.level
    group by l.level
)
select
    level,
    customer_count,
    lag(customer_count) over (order by level) as prev_level_count,
    round(100.0 * customer_count / lag(customer_count) over (order by level), 2) as pct_retained_from_prev
from survivors
order by level;
'''

pd.read_sql(query, con=engine)

,level,customer_count,prev_level_count,pct_retained_from_prev
0,1,96096,NaN,NaN
1,2,2997,96096.0,3.12
2,3,252,2997.0,8.41
3,4,49,252.0,19.44
4,5,19,49.0,38.78
5,6,11,19.0,57.89
6,7,5,11.0,45.45
7,8,2,5.0,40.00
8,9,2,2.0,100.00
9,10,1,2.0,50.00


In [10]:
query = '''
with returning_customers as (
    select customer_unique_id
    from customers
    group by customer_unique_id
    having count(*) >= 2
),
ranked_orders as (
    select
        c.customer_unique_id,
        o.order_id,
        o.order_purchase_timestamp,
        row_number() over (partition by c.customer_unique_id order by o.order_purchase_timestamp) as order_rank
    from customers c
    join orders o on c.customer_id = o.customer_id
    where c.customer_unique_id in (select customer_unique_id from returning_customers)
),
order_values as (
    select
        ro.customer_unique_id,
        ro.order_rank,
        sum(op.payment_value) as order_payment
    from ranked_orders ro
    join order_payments op on ro.order_id = op.order_id
    group by ro.customer_unique_id, ro.order_rank
)
select
    round(avg(case when order_rank = 1 then order_payment end), 2) as avg_first_order,
    round(avg(case when order_rank = 2  then order_payment end), 2) as avg_second_order,
    round(avg(case when order_rank > 1  then order_payment end), 2) as avg_subsequent_orders
from order_values;
'''

pd.read_sql(query, con=engine)

,avg_first_order,avg_second_order,avg_subsequent_orders
0,147.95,149.28,149.66


## Customer Retention Analysis

### Retention Rate
Of 99,441 customers, **96,096 (96.6%) purchased only once**. Just **3,099 customers (~3%) ever returned** for a second purchase.

Survival by purchase level:

| Purchase # | Customers | Retained from previous |
|---|---|---|
| 1 | 96,096 | — |
| 2 | 2,997 | 3.12% |
| 3 | 252 | 8.41% |
| 4 | 49 | 19.44% |
| 5+ | <20 | negligible |

The pattern: once a customer returns at all, they're more likely to return again — but the bottleneck is getting them back the first time. Only 3% clear that hurdle.

### Spend Behavior: First vs Second Orders
Among returning customers only:

| Cohort | Avg Order Value |
|---|---|
| 1st order | R$ 147.95 |
| 2nd order | R$ 149.28 |

No meaningful difference. Returning customers spend the same amount regardless of purchase number — no upsell effect, no loyalty premium.

### Business Implication
The retention problem is volume, not value. Fixing AOV for returning customers is irrelevant at this scale. The priority is increasing the 3% return rate — even moving it to 6% would double the returning customer base with zero acquisition cost.